src/programs/lsm.md

```diff
@@ -0,0 +1,191 @@
+# LSM
+
+## What is LSM
+
+LSM stands for [Linux Security Modules](https://en.wikipedia.org/wiki/Linux_Security_Modules)
+which is a framework which allows developers to write security systems on top
+of the Linux kernel. It's also briefly described in
+[the Linux kernel documentation](https://www.kernel.org/doc/html/latest/security/lsm.html).
+
+LSM is used by kernel modules or (since kernel 5.7) by eBPF programs. The most
+popular modules using LSM are AppArmor, SELinux, Smack and TOMOYO. eBPF LSM
+programs can work similar to those and let developers create their own security
+policies (or even policy systems) for Linux.
+
+The most important parts of LSM are **LSM hooks**, which are hooks placed in
+various places in the Linux kernel code. Those places are usually performing
+some actions which might be considered a security risk under some
+circumstances. But the kernel, instead of making a decision on its own,
+lets the security modules or eBPF programs to decide whether that action
+is allowed to happen or not.
+
+Those actions we are talking about are, for example:
+
+* filesystem operations
+  * opening, creating, moving and removing files
+  * mounting and unmounting filesystems
+* task/process operations
+  * allocating and freeing tasks, changing user and group identify for a task
+* socket operations
+  * creating and binding sockets
+  * receiving and sending messages
+
+Each of those actions has a corresponding LSM hook. All LSM hooks are listed in
+the [lsm_hooks.h](https://github.com/torvalds/linux/blob/master/include/linux/lsm_hooks.h)
+header inside the Linux kernel. Each LSM hook usually has some arguments with
+information about each action which should make it easier for modules/programs
+to make a decision, which are listed in the
+[lsm_hook_defs.h](https://github.com/torvalds/linux/blob/master/include/linux/lsm_hook_defs.h)
+header.
+
+Modules or eBPF programs based on LSM are triggered by those hooks and then
+they can decide whether a particular action is allowed to happen or not.
+
+For example, there is an LSM hook named `task_setnice` with the following
```

Comment:
For example consider the `task_setnice` hook, which has the following definition: